opencog/timeoctomap/TimeOctomap.cc

```diff
@@ -279,22 +288,69 @@ bool TimeOctomap::get_oldest_time_elapse_atom_observed(const opencog::Handle& at
         return false;
     }
     tl.sort();
-    if (from_d>tl.back()){
+    //check the list
+
+    if ((from_d>tl.back()) && (! is_time_point_in_range(from_d,tl.back(),curr_duration))){
         return false;
     }
     for (int i=0;i<sz;i++)
     {
        result=tl.front();
        tl.pop_front();
-       if (result>=from_d)
+       if (result>=from_d || (is_time_point_in_range(from_d,result,curr_duration)))
        {
          return true;
        }
     }
-    return false;
+    return true;
+}
+
+bool TimeOctomap::get_oldest_time_locations_atom_observed(const opencog::Handle& ato,
+                                            const time_pt& from_d,
+                                            point3d_list& result)
+{
+  time_pt tpt;
+  if (!get_oldest_time_elapse_atom_observed(ato,from_d,tpt)) return false;
+  result=get_locations_of_atom_occurence_at_time(tpt,ato);
+  //std::cout << result.size()<<"\n ";//<<from_d;
+  //std::time_t ttp=std::chrono::system_clock::to_time_t(tpt);
+  //char buff[31];
+  //strftime(buff, 30, "%Y-%m-%d %H:%M:%S ", std::localtime(&ttp));
+  //string ts(buff);
+  //std::cout << ts <<endl;
+  if (result.size()<1)return false;
+  return true;
 }
 
 bool TimeOctomap::get_last_time_elapse_atom_observed(const opencog::Handle& ato,
+                                            const time_pt& from_d,
+                                            time_pt& result)
+{
+    time_list tl=get_times_of_atom_occurence_in_map(ato);
+    //sort
+    int sz=tl.size();
+    if (sz<1)
+        return false;
+    tl.sort();
+    if ((from_d>tl.back())&&(!is_time_point_in_range(from_d,tl.back(),curr_duration)))
+        return false;
+    result=tl.back();
```

Comment:
various white-space issues in the code above.

The whole point of good white-space formatting is to make the code easier to read and to understand.  When the white-space is bad, then it takes more mental effort to understand the code.  Understanding code is already difficult, without making it harder, due to bad white-space usage.